chain/events/filter/index.go

```diff
@@ -67,34 +63,67 @@ var (
 	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
 	revertEventSeen      = `UPDATE events_seen SET reverted=true WHERE height=? AND tipset_key_cid=?`
 	restoreEventSeen     = `UPDATE events_seen SET reverted=false WHERE height=? AND tipset_key_cid=?`
 	upsertEventsSeen     = `INSERT INTO events_seen(height, tipset_key_cid, reverted) VALUES(?, ?, false) ON CONFLICT(height, tipset_key_cid) DO UPDATE SET reverted=false`
-	isTipsetProcessed    = `SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid=?`
-	getMaxHeightInIndex  = `SELECT MAX(height) FROM events_seen`
-	isHeightProcessed    = `SELECT COUNT(*) > 0 FROM events_seen WHERE height=?`
 
-	createIndexEventEmitterAddr  = `CREATE INDEX IF NOT EXISTS event_emitter_addr ON event (emitter_addr)`
+	eventExists = `SELECT MAX(id) FROM event WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	// QUERY PLAN
+	// `--SEARCH event USING INDEX event_height (height=?)
+	isTipsetProcessed = `SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid=?`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_tipset_key_cid (tipset_key_cid=?)
+	getMaxHeightInIndex = `SELECT MAX(height) FROM events_seen`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_height
+	isHeightProcessed = `SELECT COUNT(*) > 0 FROM events_seen WHERE height=?`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_height (height=?)
+
+	// When modifying indexes in this file, it is critical to test the query plan (EXPLAIN QUERY PLAN)
+	// of all the variations of queries built by prefillFilter to ensure that the query first hits
+	// an index that narrows down results to an epoch or a reasonable range of epochs. Specifically,
+	// event_tipset_key_cid or event_height should be the first index. Then further narrowing can take
+	// place within the small subset of results.
+	// Unfortunately SQLite has some quicks in index selection that mean that certain query types will
+	// bypass these indexes if alternatives are available. This has been observed specifically on
+	// queries with height ranges: `height>=X AND height<=Y`.
+	//
+	// e.g.:
+	// EXPLAIN QUERY PLAN
+	// SELECT
+	// 	event.height, event.tipset_key_cid, event_entry.indexed, event_entry.codec, event_entry.key, event_entry.value
+	// FROM event
+	// JOIN
+	// 	event_entry ON event.id=event_entry.event_id,
+	// 	event_entry ee2 ON event.id=ee2.event_id
+	// WHERE event.height>=? AND event.height<=? AND event.reverted=? AND event.emitter_addr=? AND ee2.indexed=1 AND ee2.key=?
+	// ORDER BY event.height DESC, event_entry._rowid_ ASC
+	//
+	// ->
+	//
+	// QUERY PLAN
+	// |--SEARCH event USING INDEX event_height (height>? AND height<?)
```

Comment:
I think the point was to show that the `event_height` Index does not get used if height ranges are queries. Is this the correct query plan ?

Reply:
the query plan is correct, it's updated for the current schema; I've amended the text to say:

```
	// e.g. we want to see that `event_height` is the first index used in this query:
```

it turns out that height ranges are just find if it doesn't have a better alternative. it was the composite index on `(height, reverted)` that was causing problems with ranges - it would use this for `height=? and reverted=false` but then bork on it for `height>=? and height <=? and reverted=false`!! but with a non-composite index it seems to be OK with this, we just don't get to index `reverted`.